parlai/agents/transformer/polyencoder.py

```diff
@@ -0,0 +1,267 @@
+            assert len(ctxt_tokens.shape) == 2
+            bsz = ctxt_tokens.size(0)
+        if cand_tokens is not None:
+            assert len(cand_tokens.shape) == 3
+            bsz = cand_tokens.size(0)
+            num_cands = cand_tokens.size(1)
+
+        cand_embed = None
+        ctxt_rep = None
+        ctxt_rep_mask = None
+
+        if cand_tokens is not None:
+            cand_embed = self.encoder_cand(cand_tokens.view(bsz * num_cands, -1))
+            cand_embed = cand_embed.view(bsz, num_cands, -1)
+
+        if ctxt_tokens is not None:
+            # get context_representation. Now that depends on the cases.
+            ctxt_out, ctxt_mask = self.encoder_ctxt(ctxt_tokens)
+            dim = ctxt_out.size(2)
+
+            if self.type == 'codes':
+                # Basic Attention and MultiHeadAttention share the same API.
+                ctxt_rep = self.code_attention(self.codes.repeat(bsz,1,1),
+                                               ctxt_out,
+                                               ctxt_mask)
+                ctxt_rep_mask = ctxt_rep.new_ones(bsz, self.n_codes).byte()
+
+            elif self.type == 'n_first':
+                # Expand the output if it is not long enough
+                if ctxt_out.size(1) < self.n_codes:
+                    difference = self.n_codes - ctxt_out.size(1)
+                    extra_rep = ctxt_out.new_zeros(bsz, difference, dim)
+                    ctxt_rep = torch.cat([ctxt_out, extra_rep], dim=1)
+                    extra_mask = ctxt_mask.new_zeros(bsz, difference)
+                    ctxt_rep_mask = torch.cat([ctxt_mask, extra_mask], dim=1)
+                else:
+                    ctxt_rep = ctxt_out[:, 0:self.n_codes, :]
+                    ctxt_rep_mask = ctxt_mask[:, 0:self.n_codes]
+
+        return ctxt_rep, ctxt_rep_mask, cand_embed
+
+    def score(self, ctxt_rep, ctxt_rep_mask, cand_embed):
+        """
+            Scores the candidates
+            :param ctxt_rep: 3D float tensor, bsz x ctxt_len x dim
+            :param ctxt_rep_mask: 2D byte tensor, bsz x ctxt_len, in case
+                there are some elements of the ctxt that we should not take into
+                account.
+            :param cand_embed: 3D float tensor, bsz x num_cands x dim
+
+            :returns: scores, 2D float tensor: bsz x num_cands
+        """
+
+        # reduces the context representation to a 3D tensor bsz x num_cands x dim
+        ctxt_final_rep = self.attention(cand_embed,
+                                        ctxt_rep,
+                                        ctxt_rep_mask)
+        scores = torch.sum(ctxt_final_rep * cand_embed, 2)
+        return scores
+
+    def forward(self, ctxt_tokens=None, cand_tokens=None,
+                      ctxt_rep=None, ctxt_rep_mask=None, cand_rep=None):
+        """ Due to a limitation of parlai, we have to have one single model
+            in the agent. And because we want to be able to use data-parallel,
```

Comment:
hum. different models? 
I wanted to keep one model so that all the save/load functions of Parlai would work straightforward

Reply:
ah sorry meant to say different calls to the model forward. you could factor out the score function from the forward so that you just call score instead of model(...). see how it is done here for instance: https://github.com/facebookresearch/ParlAI/blob/ba3d6204e6d740ef711d82f6e0375ee5fea4db57/parlai/agents/transformer/transformer.py#L192

Reply:
Your exemple does not need any model parameter in the _score function. _score could basically be static.

Here the score function makes use of attention (multi head attention) which adds a few parameters (query, key and value matrices) which have to be stored in the model.

Reply:
ahhh i see. ok, in this case this seems like the cleanest solution 😐

Reply:
Actually something a bit cleaner that I had tried before would be to add an argument `operation_type`
```
self.model('encode', ...)
self.model('score', ...)
```
Unfortunately this seems to bug with data-parallel when batchsize < num_gpus so I didn't find better than what's there right now